server/public/model/websocket_message.go

```diff
@@ -189,10 +190,21 @@ func (ev *WebSocketEvent) PrecomputeJSON() *WebSocketEvent {
 	return copy
 }
 
+func (ev *WebSocketEvent) RemovePrecomputedJSON() {
+	ev.precomputedJSON = nil
+}
+
 func (ev *WebSocketEvent) Add(key string, value any) {
 	ev.data[key] = value
 }
 
+// AddWithCopy copies the map and writes to a copy of that,
+// and sets the new map to the event
+func (ev *WebSocketEvent) AddWithCopy(key string, value any) {
+	ev.data = copyMap(ev.data)
+	ev.data[key] = value
+}
```

Comment:
Are these safe? From what I remember we used to avoid mutating ws events at all costs and always returned a copy when having to change something.

Reply:
Hmm .. good point, let me change that.

Reply:
Ah no, it has further problems. The issue is that we are calling this via `ShouldSendEvent` from the web_hub, but just passing the pointer. This means to modify that, we either need to assign the return value or do something like `*dst = *src`. Both seem equally invasive to me.

I tested the current code extensively in `-race` mode and didn't see any issue. I am inclined to keep it this way unless you have strong opinions?

Reply:
Okay, noting that this is safe only because we call [`PrecomputeJSON()`](https://github.com/mattermost/mattermost/blob/cc2205c6a486fdd3dd42e3a8de5a329e2ca685c4/server/channels/app/platform/web_hub.go#L489) which shallow copies the message, including the map reference, before the hook runs. Same goes for calling `RemovePrecomputedJSON()`. So extra care should be taken when using this method because it's not inherently safe.

Reply:
Yep correct.